base/docs/Docs.jl

```diff
@@ -283,15 +283,18 @@ function doc(binding::Binding, sig::Type = Union{})
         result = getdoc(resolve(binding), sig)
         result === nothing || return result
     end
-    results, groups = DocStr[], MultiDoc[]
+    results, groups, sigs = DocStr[], MultiDoc[], Any[]
     # Lookup `binding` and `sig` for matches in all modules of the docsystem.
     for mod in modules
         dict = meta(mod)
         if haskey(dict, binding)
             multidoc = dict[binding]
             push!(groups, multidoc)
             for msig in multidoc.order
-                sig <: msig && push!(results, multidoc.docs[msig])
+                if (isconcrete(sig) && sig <: msig) || (!isconcrete(sig) && msig <: sig) || sig == Union{}
```

Comment:
Maybe `!isconcrete(sig) && typeintersect(msig, sig) != Union{}` ?

Reply:
It doesn't make sense to ask if a `sig` isconcrete: we don't care if we could allocate instances of it. You could query whether it `isdispatchtuple`. But I think that would simply reduce this ternary to `typeintersect(msig, sig) != Union{}`